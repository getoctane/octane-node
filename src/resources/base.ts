import { Configuration as APIConfiguration } from '../codegen/runtime';

export abstract class BaseResource {
  private apiConfig: APIConfiguration;

  constructor(apiConfig: APIConfiguration) {
    this.apiConfig = apiConfig;
  }
}
