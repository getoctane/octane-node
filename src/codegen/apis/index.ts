/* tslint:disable */
/* eslint-disable */
export * from './AdminApi';
export * from './AvalaraApi';
export * from './BillingSettingsApi';
export * from './CouponsApi';
export * from './CreditsApi';
export * from './CustomerPortalApi';
export * from './CustomerPortalSettingsApi';
export * from './CustomersApi';
export * from './InvoicesApi';
export * from './MeasurementsApi';
export * from './MetersApi';
export * from './PaymentGatewayCredentialApi';
export * from './PricePlansApi';
export * from './RefundApi';
export * from './SelfServeSettingsApi';
export * from './WebhooksApi';
