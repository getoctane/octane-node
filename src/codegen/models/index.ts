/* tslint:disable */
/* eslint-disable */
export * from './AccountingCustomer';
export * from './AccruedRevenue';
export * from './AccruedRevenueLineItem';
export * from './ActiveSubscription';
export * from './AddAddOnsToPricePlanInput';
export * from './AddFeaturesToPricePlanInput';
export * from './AddMeteredComponentsToPricePlanInput';
export * from './AddOn';
export * from './AddOnInputArgs';
export * from './ApplyCouponInputArgs';
export * from './BankAccountInfo';
export * from './BillingCycleDate';
export * from './BillingSettings';
export * from './CardInfo';
export * from './Company';
export * from './ContactInfo';
export * from './ContactInfoInputArgs';
export * from './Coupon';
export * from './CouponInputArgs';
export * from './CreateBillingSettingsInputArgs';
export * from './CreateCreditGrantArgs';
export * from './CreateCreditTopOffPlanInputArgs';
export * from './CreateCustomerArgs';
export * from './CreateMeterArgs';
export * from './CreatePricePlanArgs';
export * from './CreateRefundArgs';
export * from './CreateSubscriptionArgs';
export * from './CreateVendorAvalaraSettingsArgs';
export * from './CreateWebhookArgs';
export * from './CreditGrant';
export * from './CreditLedger';
export * from './CreditPlan';
export * from './CreditPlanInputArgs';
export * from './CreditTopOffPlan';
export * from './Customer';
export * from './Customer1';
export * from './CustomerAvalaraSettings';
export * from './CustomerBillingSettingsInputArgs';
export * from './CustomerFeature';
export * from './CustomerLabelLimit';
export * from './CustomerMeasurementMapping';
export * from './CustomerMeasurementMappingInputArgs';
export * from './CustomerMetadata';
export * from './CustomerMetadataInput';
export * from './CustomerPaymentGatewayCredentialInputArgs';
export * from './CustomerPaymentMethodStatus';
export * from './CustomerPortalAccruedRevenue';
export * from './CustomerPortalAccruedRevenueLineItem';
export * from './CustomerPortalActiveSubscription';
export * from './CustomerPortalActiveSubscriptionInputArgs';
export * from './CustomerPortalAvailableCreditBalance';
export * from './CustomerPortalContactInfoInputArgs';
export * from './CustomerPortalCreditPurchase';
export * from './CustomerPortalDailyAccruedRevenue';
export * from './CustomerPortalInvoice';
export * from './CustomerPortalInvoiceStatus';
export * from './CustomerPortalLabelDisplayName';
export * from './CustomerPortalLabelFilter';
export * from './CustomerPortalMeter';
export * from './CustomerPortalMeterLabelFilter';
export * from './CustomerPortalMeterLabels';
export * from './CustomerPortalMeterLabelsWithDisplayName';
export * from './CustomerPortalPaymentMethod';
export * from './CustomerPortalSettings';
export * from './CustomerPortalStripeCredential';
export * from './CustomerPortalSubscription';
export * from './CustomerPortalSubscriptionInputArgs';
export * from './CustomerPortalToken';
export * from './CustomerPortalTokenInputArgs';
export * from './CustomerPortalUrl';
export * from './CustomerPortalUsage';
export * from './CustomerPortalVendor';
export * from './CustomerStatus';
export * from './CustomerTag';
export * from './CustomerUsage';
export * from './CycleUsage';
export * from './DailyAccruedRevenue';
export * from './DailyUsage';
export * from './DeleteDiscountInputArgs';
export * from './DeleteSubscriptionArgs';
export * from './Discount';
export * from './DiscountInputArgs';
export * from './EntityUseCode';
export * from './Feature';
export * from './FeatureInputArgs';
export * from './Invoice';
export * from './Limit';
export * from './LimitInputArgs';
export * from './LineItems';
export * from './ListCreditGrants';
export * from './ListCreditGrantsArgs';
export * from './ListPricePlans';
export * from './Measurement';
export * from './Meter';
export * from './MeteredComponent';
export * from './MeteredComponentInputArgs';
export * from './MeteredComponentLabelLimit';
export * from './MeteredComponentLabelLimitInputArgs';
export * from './ModelError';
export * from './PastInvoice';
export * from './PastInvoices';
export * from './PaymentGatewayCredential';
export * from './PaymentGatewayCredentialInputArgs';
export * from './PriceInputArgs';
export * from './PriceInputArgs1';
export * from './PricePlan';
export * from './PricePlan1';
export * from './PricePlanTag';
export * from './PriceScheme';
export * from './PriceSchemeInputArgs';
export * from './PriceTier';
export * from './Refund';
export * from './RemoveFeaturesFromPricePlanInput';
export * from './Retry';
export * from './RevenueBreakdown';
export * from './RevenueRecognitionEntry';
export * from './RevenueRecognitionInput';
export * from './RevenueResponse';
export * from './RollApiKeyArgs';
export * from './RollApiKeys';
export * from './SelfServeCustomization';
export * from './SelfServePricePlansInputArgs';
export * from './SelfServeSettings';
export * from './Subscription';
export * from './SubscriptionAddOnInput';
export * from './SubscriptionAddOnItem';
export * from './TaxCode';
export * from './Trial';
export * from './TrialInputArgs';
export * from './UpdateBillingSettingsInputArgs';
export * from './UpdateCreditTopOffPlanInputArgs';
export * from './UpdateCustomerArgs';
export * from './UpdateCustomerAvalaraSettingsArgs';
export * from './UpdateCustomerPortalSettingsInputArgs';
export * from './UpdateInvoiceStatusArgs';
export * from './UpdateMeterArgs';
export * from './UpdateMeteredComponentLimitsInput';
export * from './UpdatePricePlanArgs';
export * from './UpdatePricePlanInPlaceArgs';
export * from './UpdateSelfServeSettingsArgs';
export * from './UpdateSubscriptionArgs';
export * from './UpdateSubscriptionInPlaceArgs';
export * from './UpdateVendorAvalaraSettingsArgs';
export * from './ValidateAddressResp';
export * from './ValidateCredentialsArgs';
export * from './ValidateCredentialsResp';
export * from './VendorAvalaraSettings';
export * from './Webhook';
