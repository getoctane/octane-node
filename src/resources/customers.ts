import {
  CreateCustomerArgs,
  CreateSubscriptionArgs,
  Customer,
  CustomerFeature,
  CustomerPaymentGatewayCredentialInputArgs,
  CustomersApi,
  DeleteSubscriptionArgs,
  PaymentGatewayCredential,
  Subscription,
  UpdateCustomerArgs,
  UpdateSubscriptionArgs,
} from '../codegen/api';
import { ClientConfiguration } from '../types';
import { Configuration as APIConfiguration } from '../codegen/configuration';
import { BaseResource } from './base';

class Customers extends BaseResource {
  private api: CustomersApi;

  constructor(apiConfig: APIConfiguration, clientConfig: ClientConfiguration) {
    super(clientConfig);
    this.api = new CustomersApi(apiConfig);
  }

  /**
   * Create a new customer.
   */
  public create(body: CreateCustomerArgs, options?: any): Promise<Customer> {
    return this.api.customersPost(body, options).then(this.formatResponse);
  }

  /**
   * Fetch a customer by their unique name.
   */
  public retrieve(customerName: string, options?: any): Promise<Customer> {
    return this.api
      .customersCustomerNameGet(customerName, options)
      .then(this.formatResponse);
  }

  /**
   * Update a single customer.
   */
  public update(
    customerName: string,
    body: UpdateCustomerArgs,
    options?: any,
  ): Promise<Customer> {
    // NOTE: order or arguments switched here
    return this.api
      .customersCustomerNamePut(body, customerName, options)
      .then(this.formatResponse);
  }

  /**
   * Retrieve all customers for a given vendor.
   */
  public list(options?: any): Promise<Customer[]> {
    return this.api.customersGet(options).then(this.formatResponse);
  }

  /**
   * Delete a customer by their unique name.
   */
  public delete(customerName: string, options?: any): Promise<Response> {
    // TODO: void the response as it is inconsistent with others
    // NOTE: there's no need to format the response from a delete endpoint
    return this.api.customersCustomerNameDelete(customerName, options);
  }

  /**
   * Add PaymentGatewayCredential for a customer.
   */
  public createPaymentGatewayCredential(
    customerName: string,
    body: CustomerPaymentGatewayCredentialInputArgs,
    options?: any,
  ): Promise<PaymentGatewayCredential> {
    // NOTE: order or arguments switched here
    return this.api
      .customersCustomerNamePaymentGatewayCredentialsPost(
        body,
        customerName,
        options,
      )
      .then(this.formatResponse);
  }

  /**
   * Create a new subscription for a customer / price plan combination (by unique name).
   */
  public createSubscription(
    customerName: string,
    body: CreateSubscriptionArgs,
    options?: any,
  ): Promise<Subscription> {
    // NOTE: order or arguments switched here
    return this.api
      .customersCustomerNameSubscriptionsPost(body, customerName, options)
      .then(this.formatResponse);
  }

  /**
   * Get all subscriptions for the customer.
   */
  public listSubscriptions(
    customerName: string,
    options?: any,
  ): Promise<Subscription[]> {
    return this.api
      .customersCustomerNameSubscriptionsGet(customerName, options)
      .then(this.formatResponse);
  }

  /**
   * Update a subscription for a specific customer (by customer name).
   */
  public updateSubscription(
    customerName: string,
    body: UpdateSubscriptionArgs,
    options?: any,
  ): Promise<Subscription> {
    // NOTE: order or arguments switched here
    return this.api
      .customersCustomerNameSubscriptionPut(body, customerName, options)
      .then(this.formatResponse);
  }

  /**
   * Delete a subscription for a specific customer (by customer name).
   */
  public deleteSubscription(
    customerName: string,
    body: DeleteSubscriptionArgs,
    options?: any,
  ): Promise<Response> {
    // TODO: void the response as it is inconsistent with others
    // NOTE: order or arguments switched here
    return this.api
      .customersCustomerNameSubscriptionDelete(body, customerName, options)
      .then(this.formatResponse);
  }

  /**
   * Retreive a customer's access to feature/limitation.
   */
  public retrieveFeature(
    customerName: string,
    featureName: string,
    options?: any,
  ): Promise<CustomerFeature> {
    // TODO: void the response as it is inconsistent with others
    // NOTE: order or arguments switched here
    return this.api
      .customersCustomerNameFeaturesFeatureNameGet(
        featureName,
        customerName,
        options,
      )
      .then(this.formatResponse);
  }
}

export { Customers };
