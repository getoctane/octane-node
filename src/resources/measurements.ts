import { Measurement, MeasurementsApi } from '../codegen/api';
import { Configuration as APIConfiguration } from '../codegen/configuration';

class Measurements {
  private api: MeasurementsApi;

  constructor(apiConfig: APIConfiguration) {
    this.api = new MeasurementsApi(apiConfig);
  }

  /**
   * Send Measurement (one or multiple).
   */
  public create(body: Measurement | Array<Measurement>, options?: any) {
    if (body instanceof Array) {
      return this.api.measurementsMultiPost(body, options);
    }
    return this.api.measurementsPost(body, options);
  }
}

export { Measurements };
