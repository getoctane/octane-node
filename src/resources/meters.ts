import { MeterInputArgs, MetersApi, UpdateMeterArgs } from '../codegen/api';
import { Configuration as APIConfiguration } from '../codegen/configuration';

class Meters {
  private api: MetersApi;

  constructor(apiConfig: APIConfiguration) {
    this.api = new MetersApi(apiConfig);
  }

  /**
   * Create a new Meter.
   */
  public create(body: MeterInputArgs, options?: any) {
    return this.api.metersPost(body, options);
  }

  /**
   * Fetch a meter by its unique name.
   */
  public retrieve(meterName: string, options?: any) {
    return this.api.metersMeterNameGet(meterName, options);
  }

  /**
   * Update a meter by its unique name.
   */
  public update(meterName: string, body: UpdateMeterArgs, options?: any) {
    // NOTE: order or arguments switched here
    return this.api.metersMeterNamePut(body, meterName, options);
  }

  /**
   * Retrieve all meters for a given vendor.
   */
  public list(options?: any) {
    return this.api.metersGet(options);
  }

  /**
   * Delete a meter by its unique name.
   */
  public delete(meterName: string, options?: any) {
    // TODO: void the response as it is inconsistent with others
    return this.api.metersMeterNameDelete(meterName, options);
  }
}

export { Meters };
