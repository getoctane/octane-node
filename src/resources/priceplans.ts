import {
  PricePlansApi,
  CreatePricePlanArgs,
  UpdatePricePlanArgs,
} from '../codegen/api';
import { Configuration as APIConfiguration } from '../codegen/configuration';

class PricePlans {
  private api: PricePlansApi;

  constructor(apiConfig: APIConfiguration) {
    this.api = new PricePlansApi(apiConfig);
  }

  /**
   * Create Price Plan.
   */
  public create(body: CreatePricePlanArgs, options?: any) {
    return this.api.pricePlansPost(body, options);
  }

  /**
   * Fetch an existing price plan.
   */
  public retrieve(pricePlanName: string, options?: any) {
    return this.api.pricePlansPricePlanNameGet(pricePlanName, options);
  }

  /**
   * Update a meter by its unique name.
   */
  public update(
    pricePlanName: string,
    body: UpdatePricePlanArgs,
    options?: any,
  ) {
    // NOTE: order or arguments switched here
    return this.api.pricePlansPricePlanNamePut(body, pricePlanName, options);
  }

  /**
   * Retrieve all meters for a given vendor.
   */
  public list(options?: any) {
    return this.api.pricePlansGet(options);
  }

  /**
   * Delete a meter by its unique name.
   */
  public delete(pricePlanName: string, options?: any) {
    // TODO: void the response as it is inconsistent with others
    return this.api.pricePlansPricePlanNameDelete(pricePlanName, options);
  }
}

export { PricePlans };
