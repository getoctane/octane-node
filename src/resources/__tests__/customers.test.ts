import { Customers } from 'resources/customers';
import fetchMock from 'jest-fetch-mock';
import { Configuration, CreateCustomerArgs } from 'codegen';

const API_CONFIG = new Configuration({
  accessToken: '867-5309',
  basePath: 'https://api.made.up.host:1337',
});

const FAKE_CUSTOMER: CreateCustomerArgs = {
  name: 'chuck_testa',
  displayName: 'Chuck Testa',
  contactInfo: {
    addressLine1: '1600 Pennsylvania Avenue NW',
    addressLine2: 'Apt. 4F',
    city: 'Washington',
    state: 'DC',
    zipcode: '20500',
  },
};

// The response comes back from the server in snake_case.
// This response is just FAKE_CUSTOMER but in snake_case.
const FAKE_RESPONSE = {
  name: 'chuck_testa',
  display_name: 'Chuck Testa',
  contact_info: {
    address_line_1: '1600 Pennsylvania Avenue NW',
    address_line_2: 'Apt. 4F',
    city: 'Washington',
    state: 'DC',
    zipcode: '20500',
  },
};

describe('Customers resource', () => {
  it('makes create requests', async () => {
    const c = new Customers(API_CONFIG);
    fetchMock.once(JSON.stringify(FAKE_RESPONSE));
    expect(await c.create(FAKE_CUSTOMER)).toEqual(FAKE_CUSTOMER);
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('makes retrieve requests', async () => {
    const c = new Customers(API_CONFIG);
    fetchMock.once(JSON.stringify(FAKE_RESPONSE));
    expect(await c.retrieve('chuck_testa')).toEqual(FAKE_CUSTOMER);
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('makes update requests', async () => {
    const c = new Customers(API_CONFIG);
    fetchMock.once(JSON.stringify(FAKE_RESPONSE));
    expect(await c.update('chuck_testa', FAKE_CUSTOMER)).toEqual(FAKE_CUSTOMER);
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('makes list requests', async () => {
    const c = new Customers(API_CONFIG);
    fetchMock.once(JSON.stringify([FAKE_RESPONSE]));
    expect(await c.list()).toEqual([FAKE_CUSTOMER]);
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('makes delete requests', async () => {
    const c = new Customers(API_CONFIG);
    fetchMock.once(JSON.stringify(FAKE_CUSTOMER));
    expect(await c.delete('chuck_testa')).toBeUndefined();
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });
});
