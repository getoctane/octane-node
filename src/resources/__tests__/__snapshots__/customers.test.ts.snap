// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customers resource makes create requests 1`] = `
[
  [
    "https://api.made.up.host:1337/customers/",
    {
      "body": "{"contact_info":{"address_line_2":"Apt. 4F","address_line_1":"1600 Pennsylvania Avenue NW","zipcode":"20500","city":"Washington","state":"DC"},"name":"chuck_testa","display_name":"Chuck Testa"}",
      "credentials": undefined,
      "headers": {
        "Authorization": "Bearer 867-5309",
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`Customers resource makes delete requests 1`] = `
[
  [
    "https://api.made.up.host:1337/customers/chuck_testa",
    {
      "body": undefined,
      "credentials": undefined,
      "headers": {
        "Authorization": "Bearer 867-5309",
      },
      "method": "DELETE",
    },
  ],
]
`;

exports[`Customers resource makes list requests 1`] = `
[
  [
    "https://api.made.up.host:1337/customers/",
    {
      "body": undefined,
      "credentials": undefined,
      "headers": {
        "Authorization": "Bearer 867-5309",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Customers resource makes retrieve requests 1`] = `
[
  [
    "https://api.made.up.host:1337/customers/chuck_testa",
    {
      "body": undefined,
      "credentials": undefined,
      "headers": {
        "Authorization": "Bearer 867-5309",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Customers resource makes update requests 1`] = `
[
  [
    "https://api.made.up.host:1337/customers/chuck_testa",
    {
      "body": "{"contact_info":{"address_line_2":"Apt. 4F","address_line_1":"1600 Pennsylvania Avenue NW","zipcode":"20500","city":"Washington","state":"DC"},"name":"chuck_testa","display_name":"Chuck Testa"}",
      "credentials": undefined,
      "headers": {
        "Authorization": "Bearer 867-5309",
        "Content-Type": "application/json",
      },
      "method": "PUT",
    },
  ],
]
`;
